packages/bundlers/default/src/DefaultBundler.js

```diff
@@ -38,385 +50,712 @@ const HTTP_OPTIONS = {
-                (dependency.priority === 'parallel' &&
-                  !dependency.needsStableName)
-                  ? false
-                  : parentBundle.needsStableName,
-              bundleBehavior: dependency.bundleBehavior ?? asset.bundleBehavior,
-              isSplittable: asset.isBundleSplittable ?? true,
-              pipeline: asset.pipeline,
-            });
-            bundleByType.set(bundle.type, bundle);
-            bundlesByEntryAsset.set(asset, bundle);
-            bundleGraph.createAssetReference(dependency, asset, bundle);
+  // Models bundleRoots and the assets that require it synchronously
+  let reachableRoots: ContentGraph<Asset> = new ContentGraph();
+  for (let [root] of bundleRoots) {
+    let rootNodeId = reachableRoots.addNodeByContentKeyIfNeeded(root.id, root);
+    assetGraph.traverse((node, isAsync, actions) => {
+      if (node.value === root) {
+        return;
+      }
 
-            // The bundle may have already been created, and the graph gave us back the original one...
-            if (!bundleRoots.has(bundle)) {
-              bundleRoots.set(bundle, [asset]);
+      if (node.type === 'dependency') {
+        let dependency = node.value;
+        if (dependencyBundleGraph.hasContentKey(dependency.id)) {
+          if (dependency.priority === 'lazy') {
+            let assets = assetGraph.getDependencyAssets(dependency);
+            if (assets.length === 0) {
+              return node;
+            }
+
+            invariant(assets.length === 1);
+            let bundleRoot = assets[0];
+            if (dependency.specifierType !== 'url') {
+              asyncBundleRootGraph.addEdge(
+                asyncBundleRootGraph.getNodeIdByContentKey(root.id),
+                asyncBundleRootGraph.getNodeIdByContentKey(bundleRoot.id),
+              );
             }
           }
+          actions.skipChildren();
+          return;
         }
+        return;
+      }
 
-        return {
-          ...context,
-          parentNode: node,
-        };
-      },
-    });
+      let nodeId = reachableRoots.addNodeByContentKeyIfNeeded(
+        node.value.id,
+        node.value,
+      );
+      reachableRoots.addEdge(rootNodeId, nodeId);
+    }, root);
+  }
 
-    for (let [bundle, rootAssets] of bundleRoots) {
-      for (let asset of rootAssets) {
-        bundleGraph.addEntryToBundle(asset, bundle);
+  // Maps a given bundleRoot to the bundleRoots reachable from it,
+  // and the assets reachable from each of these bundleRoots
+  let ancestorAssets: Map<
+    BundleRoot,
+    Map<BundleRoot, Array<Asset> | null>,
+  > = new Map();
+
+  // Reference count of each asset available within a given bundleRoot's bundle group
+  let assetRefsInBundleGroup: DefaultMap<
+    BundleRoot,
+    DefaultMap<Asset, number>,
+  > = new DefaultMap(() => new DefaultMap(() => 0));
+
+  // Step 4: Determine assets that should be duplicated by computing asset availability in each bundle group
```

Comment:
Step 4 is our answer to deduplication, or rather, removing the need for deduplication. In this step, we precompute the information we need for a bundleRoot, at that node's parent. (this "information" is called ancestorAssets, and it is the total set of assets available to a given bundle at that bundles load time)

The ancestorAsset information is modified in two parts. 
1. Consider assets available from within a node's bundleGroup first, and modify that node's available assets.
2. Peek ahead at this node's children and modify their availableAssets with what we gathered in part 1. (By intersecting)

This effectively replaces upwards traversal and nested traversals, since we "peek ahead" and modify a node's ancestorAssets by intersection, meaning we have 1 traversal.